File: client/java/src/main/java/com/be/mesh/client/prsim/Tokenizer.java
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */
package com.be.mesh.client.prsim;

import com.be.mesh.client.annotate.Index;
import com.be.mesh.client.annotate.MPI;
import com.be.mesh.client.annotate.SPI;
import com.be.mesh.client.struct.*;

import java.time.Duration;

/**
 * <pre>
 *      +--------+                               +---------------+
 *      |        |--(A)- Authorization Request ->|   Resource    |
 *      |        |                               |     Owner     |
 *      |        |<-(B)-- Authorization Grant ---|               |
 *      |        |                               +---------------+
 *      |        |
 *      |        |                               +---------------+
 *      |        |--(C)-- Authorization Grant -->| Authorization |
 *      | Client |                               |     Server    |
 *      |        |<-(D)----- Access Token -------|               |
 *      |        |                               +---------------+
 *      |        |
 *      |        |                               +---------------+
 *      |        |--(E)----- Access Token ------>|    Resource   |
 *      |        |                               |     Server    |
 *      |        |<-(F)--- Protected Resource ---|               |
 *      +--------+                               +---------------+
 * </pre>
 *
 * @author coyzeng@gmail.com
 */
@SPI("mesh")
public interface Tokenizer {

    /**
     * Apply a node token.
     *
     * @param kind     token kind
     * @param duration max life cycle
     * @return node token
     */
    @MPI("mesh.trust.apply")
    String apply(@Index(0) String kind, @Index(1) Duration duration);

    /**
     * Verify some token verifiable.
     *
     * @param token node token
     * @return true if verify passed
     */
    @MPI("mesh.trust.verify")
    boolean verify(@Index(0) String token);

    /**
     * OAuth2 quick authorize, contains grant code and code authorize.
     */
    @MPI("mesh.oauth2.quickauth")
    AccessToken quickauth(@Index(0) Credential credential);

    /**
     * OAuth2 code grant.
     */
    @MPI("mesh.oauth2.grant")
    AccessGrant grant(@Index(0) Credential credential);

    /**
     * OAuth2 accept grant code.
     */
    @MPI("mesh.oauth2.accept")
    AccessCode accept(@Index(0) String code);

    /**
     * OAuth2 reject grant code.
     */
    @MPI("mesh.oauth2.accept")
    void reject(@Index(0) String code);

    /**
     * OAuth2 code authorize.
     */
    @MPI("mesh.oauth2.authorize")
    AccessToken authorize(@Index(0) String code);

    /**
     * OAuth2 authenticate.
     */
    @MPI("mesh.oauth2.authenticate")
    AccessID authenticate(@Index(0) String token);

    /**
     * OAuth2 auth token refresh.
     */
    @MPI("mesh.oauth2.refresh")
    AccessToken refresh(@Index(0) String token);
}

File: plugin/prsim/tokenizer.go
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

package prsim

import (
	"context"
	"github.com/be-io/mesh/client/golang/cause"
	"github.com/be-io/mesh/client/golang/prsim"
	"github.com/be-io/mesh/client/golang/system"
	"github.com/be-io/mesh/client/golang/tool"
	"github.com/be-io/mesh/client/golang/types"
	"time"
)

var _ prsim.Tokenizer = new(PRSITokenizer)

// PRSITokenizer
// @MPS(macro.MeshMPS)
// @SPI(macro.MeshSPI)
type PRSITokenizer struct {
	Name  string
	Kind  string
	Token string
}

func (that *PRSITokenizer) Apply(ctx context.Context, kind string, duration time.Duration) (string, error) {
	token := tool.NextID()
	if err := system.PutWithCache(ctx, aware.Cache, token, token, duration); nil != err {
		return "", cause.Error(err)
	}
	return token, nil
}

func (that *PRSITokenizer) Verify(ctx context.Context, token string) (bool, error) {
	entity, err := aware.Cache.Get(ctx, token)
	if nil != err {
		return false, cause.Error(err)
	}
	if err := aware.Cache.Remove(ctx, token); nil != err {
		return false, cause.Error(err)
	}
	return nil != entity, nil
}

func (that *PRSITokenizer) Quickauth(ctx context.Context, credential *types.Credential) (*types.AccessToken, error) {
	return aware.Tokenizer.Quickauth(ctx, credential)
}

func (that *PRSITokenizer) Grant(ctx context.Context, credential *types.Credential) (*types.AccessGrant, error) {
	return aware.Tokenizer.Grant(ctx, credential)
}

func (that *PRSITokenizer) Accept(ctx context.Context, code string) (*types.AccessCode, error) {
	return aware.Tokenizer.Accept(ctx, code)
}

func (that *PRSITokenizer) Reject(ctx context.Context, code string) error {
	return aware.Tokenizer.Reject(ctx, code)
}

func (that *PRSITokenizer) Authorize(ctx context.Context, code string) (*types.AccessToken, error) {
	return aware.Tokenizer.Authorize(ctx, code)
}

func (that *PRSITokenizer) Authenticate(ctx context.Context, token string) (*types.AccessID, error) {
	return aware.Tokenizer.Authenticate(ctx, token)
}

func (that *PRSITokenizer) Refresh(ctx context.Context, token string) (*types.AccessToken, error) {
	return aware.Tokenizer.Refresh(ctx, token)
}
